lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -228,25 +257,223 @@ void DominatedAccessRemoval::optimize() {
   if (func.empty())
-    return;
+    return false;
 
   if (!analyze()) {
     LLVM_DEBUG(llvm::dbgs()
                << "Bailed on function: " << func.getName() << "\n");
-    return;
+    return false;
   }
   optimize();
+  return true;
+}
+
+namespace {
+class LoopDominatingAccessAdder {
+  SmallVector<SILLoop *, 8> bottomUpWorkList;
+  DominanceInfo *domInfo;
+
+public:
+  LoopDominatingAccessAdder(SILLoop *topLevelLoop, DominanceInfo *domInfo)
+      : domInfo(domInfo) {
+    // Collect loops for a recursive bottom-up traversal in the loop tree.
+    bottomUpWorkList.push_back(topLevelLoop);
+    for (unsigned i = 0; i < bottomUpWorkList.size(); ++i) {
+      auto *loop = bottomUpWorkList[i];
+      for (auto *subLoop : *loop) {
+        bottomUpWorkList.push_back(subLoop);
+      }
+    }
+  }
+
+  void perform();
+
+protected:
+  /// \brief Collect a set of instructions that can be dominated
+  void
+  analyzeCurrentLoop(SILLoop *currnetLoop,
+                     SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec);
+
+  /// \brief Optimize the current loop nest.
+  void optimizeLoop(SILLoop *currnetLoop,
+                    SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec);
+};
+} // namespace
+
+void LoopDominatingAccessAdder::analyzeCurrentLoop(
+    SILLoop *currnetLoop,
+    SmallVectorImpl<BeginAccessInst *> &dominatableInstrVec) {
+  auto *preheader = currnetLoop->getLoopPreheader();
+  if (!preheader) {
+    // Can't dominate this loop - bail
+    return;
+  }
+  for (auto *block : currnetLoop->getBlocks()) {
+    for (auto &instr : *block) {
+      auto *beginAccess = dyn_cast<BeginAccessInst>(&instr);
+      if (!beginAccess) {
+        continue;
+      }
+      if (beginAccess->getEnforcement() != SILAccessEnforcement::Dynamic) {
+        continue;
+      }
+      if (!beginAccess->hasNoNestedConflict()) {
+        continue;
+      }
+      auto operand = beginAccess->getOperand();
+      if (auto *defInstr = operand->getDefiningInstruction()) {
+        if (currnetLoop->contains(defInstr->getParent())) {
+          // Loop variant operand
+          continue;
+        }
+      } else {
+        auto *defArg = dyn_cast<SILArgument>(operand);
+        assert(defArg && "Expected an argument for operand");
+        if (currnetLoop->contains(defArg->getParent())) {
+          // Loop varaint argument
+          continue;
+        }
+      }
+      // We can (potentially) dominate this access outside of the loop,
+      // add it to dominatableInstrVec.
+      // optimizeLoop phase will do final checks,
+      // if it currently support this dominatable access, it will add a
+      // dominating access outside of the loop
+      dominatableInstrVec.push_back(beginAccess);
+    }
+  }
+}
+
+static bool skipAccess(DominanceInfo *domInfo, SILValue &operand,
+                       SILBasicBlock *preheader, TermInst *preheaderTerm) {
+  if (auto *defInstr = operand->getDefiningInstruction()) {
```

Comment:
You can replace this whole function with a call to `getParentBlock` followed by a single domanince check between the operand's block and the preheader block. Also, I *think* this dominance check should be an assert, if I understand the preconditions.